File: problem_023.rs
// -*- compile-command: "rustc -o problem_023_rs problem_023.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn main() {
    println!("Project Euler - Problem 23:");
    println!("Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.\n");

    const LIMIT: usize = 28123; // Given as largest abundant sum

    // Find sums of all numbers less than LIMIT
    // and Isolate abundant numbers into vector
    let mut sums: [usize; LIMIT] = [0; LIMIT];
    let mut abundant = Vec::new();
    for n in 1..LIMIT {
        for m in 2..LIMIT {
            let t = n * m;
            if t > LIMIT {
                break;
            }
            sums[t] += n;
        }
        if sums[n] > n{
            abundant.push(n);
        }
    }

    // Add each pair of abundant numbers, track in table of bools
    let mut abundant_sum: [bool; LIMIT] = [false; LIMIT];
    for a in 0..abundant.len() {
        for b in a..abundant.len() {
            let n = abundant[a] + abundant[b];
            if n >= LIMIT {
                break;
            }
            abundant_sum[n] = true;
        }
    }

    // Sum all the numbers that are not a sum of 2 abundant numbers (false)
    let mut sum = 0;
    for n in 0..(LIMIT-1) {
        if abundant_sum[n] == false {
            sum += n;
        }
    }

    println!("{}", sum);
}

File: problem_023.js
#!/usr/bin/node

console.log("Project Euler - Problem 23:");
console.log("Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.\n");

const limit = 28123;
var numbers = [];
var abundant_numbers = [];

// Initialize array
for (var n=0; n<limit; n++)
    numbers.push(1);

// Find all divisor sums
for (var n=2; n<limit; n++){
    let i = n + n;
    while (i < limit) {
        numbers[i] += n;
        i += n;
    }

    // Keep track of abundant numbers
    if (numbers[n] > n)
        abundant_numbers.push(n);
}

// Re-use array for abundant sums
for (var n=0; n<limit; n++)
    numbers[n] = 0;


// Add all the abundant numbers together
for (var a=0; a<abundant_numbers.length; a++)
    for (var b=a; b<abundant_numbers.length; b++){
	let abundant_sum = abundant_numbers[a] + abundant_numbers[b];
	if (abundant_sum > limit) break;
	numbers[abundant_sum] += 1;
    }

// Finally, sum all the numbers we managed not to create using abundant sums
var sum = 0;
for (var n=0; n<limit; n++)
    if (numbers[n] == 0)
	sum += n;

console.log(sum);
